app/uk/gov/hmrc/pensionschemereturn/utils/FutureUtils.scala

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.pensionschemereturn.utils
+
+import scala.concurrent.{ExecutionContext, Future}
+
+object FutureUtils {
+
+  implicit class FutureOps[A](val future: Future[A]) extends AnyVal {
+
+    def tap[B](f: A => Future[B])(implicit ec: ExecutionContext): Future[A] =
+      future.flatMap(a => f(a).as(a))
+
+    def tapError[B](f: Throwable => Future[B])(implicit ec: ExecutionContext): Future[A] =
```

Comment:
In my experience tap is used to execute a side effect while retaining the result of the effect (in this context it would almost be like peeking into the future and doing something with that result but returning the original result) - this function seems more like a `flatMapError` as we are changing the result

for reference this is `tap` in the scala standard lib - https://www.scala-lang.org/api/2.13.6/scala/util/ChainingOps.html#tap[U](f:A=%3EU):A

I know that these are just helper functions to cut down on syntax so if no one has a problem with the naming, I'm happy with it

Reply:
Neither of these methods change the original result although I have noticed a bug in tap just now where if `f` fails then the result would be changed.
For `tapError` if `f` succeeds we return a failed future of the original throwable and if `f` fails we again return a failed future of the original throwable  unless I'm missing something 

flatMapError would have an implementation of
```
def flatMapError[B](f: Throwable => Future[B])(implicit ec: ExecutionContext): Future[B] =
  future.recoverWith {
    case t => f(t)
  }
```